sqlair.go

```diff
@@ -152,19 +256,29 @@ func (q *Query) Iter() *Iterator {
 	if q.err != nil {
 		return &Iterator{err: q.err}
 	}
+	if q.tx != nil && q.tx.isDone() {
+		return &Iterator{err: ErrTXDone}
+	}
 	var result sql.Result
 	var rows *sql.Rows
 	var err error
 	var cols []string
 	if q.qe.HasOutputs() {
-		rows, err = q.qs.QueryContext(q.ctx, q.qe.QuerySQL(), q.qe.QueryArgs()...)
+		rows, err = q.sqlstmt.QueryContext(q.ctx, q.qe.QueryArgs()...)
 		if err == nil { // if err IS nil
 			cols, err = rows.Columns()
 		}
 	} else {
-		result, err = q.qs.ExecContext(q.ctx, q.qe.QuerySQL(), q.qe.QueryArgs()...)
+		result, err = q.sqlstmt.ExecContext(q.ctx, q.qe.QueryArgs()...)
+	}
+	if err != nil {
+		return &Iterator{qe: q.qe, err: err}
 	}
-	return &Iterator{qe: q.qe, rows: rows, cols: cols, err: err, result: result}
+	var sqlstmt *sql.Stmt
+	if q.tx != nil {
+		sqlstmt = q.sqlstmt
```

Comment:
Sorry for missing this earlier, but there's something here that doesn't feel quite right.  This is handing to Iter a sqlstmt for being closed, but that sqlstmt lives in the Query. That can't be right. There's nothing in the API of the Query saying that it is a requirement for Iter.Close to be called directly or via one of the other methods, for example,  and that this is a requirement for resources not to leak. Indeed it would be unfriendly to ask for that.  Then, what if Iter is called a second time in the Query?  This woud find a closed sqlstmt, and break down?

Luckily, there's a trivial solution which will actually simplify other things too: create the transaction-bound stmt inside this function instead of elsewhere, and close it immediately after this function is done with it. Then let's clean up the rest of the code since we don't need to track it around anymore.

Reply:
If the connection is closed before the query results have been read the first call to `Rows.Next` returns false so sadly the quick fix doesn't work.

You are right that it would be hostile to require the user to call `Close` though. The `database/sql` rows object will automatically close when `Rows.Next` returns false. I've added this behavior to the `Iter` so that the connection is closed when `Iter.Next` returns false.

Reply:
Why would the connection be closed?  The connection is associated with the TX. If people close the transaction (Commit, Rollback) before reading results are finished, they will get an error, as they should?

> I've added this behavior to the Iter so that the connection is closed when Iter.Next returns false.

How would it make sense to close the connection because the iterator is finished?

Reply:
Ah sorry, I meant to write prepared statement, not connection.

Reply:
The code I've implemented closes the stmt, not the connection.